python/tvm/contrib/coreml.py

```diff
@@ -0,0 +1,230 @@
+# Licensed to the Apache Software Foundation (ASF) under one
+# or more contributor license agreements.  See the NOTICE file
+# distributed with this work for additional information
+# regarding copyright ownership.  The ASF licenses this file
+# to you under the Apache License, Version 2.0 (the
+# "License"); you may not use this file except in compliance
+# with the License.  You may obtain a copy of the License at
+#
+#   http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing,
+# software distributed under the License is distributed on an
+# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+# KIND, either express or implied.  See the License for the
+# specific language governing permissions and limitations
+# under the License.
+# pylint: disable=invalid-name, unused-argument, import-outside-toplevel
+"""Utility to compile CoreML models"""
+
+import os
+import shutil
+
+import tvm._ffi
+from ..relay.expr_functor import ExprVisitor
+from ..relay.expr import Constant
+from . import xcode, coreml_runtime
+
+def _convert_add(builder, name, inputs, outputs, args, attrs):
+    builder.add_elementwise(
+        name=name,
+        input_names=inputs,
+        output_name=outputs[0],
+        mode='ADD'
+    )
+
+def _convert_multiply(builder, name, inputs, outputs, args, attrs):
+    builder.add_elementwise(
+        name=name,
+        input_names=inputs,
+        output_name=outputs[0],
+        mode='MULTIPLY'
+    )
+
+def _convert_clip(builder, name, inputs, outputs, args, attrs):
+    builder.add_clip(
+        name=name,
+        input_name=inputs[0],
+        output_name=outputs[0],
+        min_value=attrs.a_min,
+        max_value=attrs.a_max
+    )
+
+def _convert_batch_flatten(builder, name, inputs, outputs, args, attrs):
+    builder.add_flatten_to_2d(
+        name=name,
+        input_name=inputs[0],
+        output_name=outputs[0]
+    )
+
+def _convert_softmax(builder, name, inputs, outputs, args, attrs):
+    builder.add_softmax_nd(
+        name=name,
+        input_name=inputs[0],
+        output_name=outputs[0],
+        axis=int(attrs['axis'])
+    )
+
+def _convert_conv2d(builder, name, inputs, outputs, args, attrs):
+    assert isinstance(args[1], Constant)
+    assert attrs['kernel_layout'] in ['HWIO', 'OIHW']
```

Comment:
These sort of asserts might be a natural fit for a 'check' function in the graph annotator, so unsupported variants aren't offloaded in the first place.